runtime/strings/strings.cpp

```diff
@@ -0,0 +1,147 @@
+#include<assert.h>
+#include<stdlib.h>
+#include<iostream>
+#include<cstdint>
+
+extern "C" {
+
+struct blockheader {
+	int64_t len;
+};
+
+struct string {
+	blockheader b;
+	char data[0];
+};
+
+bool hook_STRINGS_gt(const string * a, const string * b) {
+    for (int i = 0; i < a->b.len; ++i) {
+        if (a->data[i] < b->data[i]) return false;
+        if (a->data[i] > b->data[i]) return true;
+        if (i >= b->b.len)  return true;
+    }
+    return false;
+}
+
+bool hook_STRINGS_gte(const string * a, const string * b) {
+    for (int i = 0; i < a->b.len; ++i) {
+        if (a->data[i] < b->data[i]) return false;
+        if (a->data[i] > b->data[i]) return true;
+        if (i >= b->b.len)  return true;
+    }
+    return a->b.len == b->b.len;
+}
+
+bool hook_STRINGS_lt(const string * a, const string * b) {
+    for (int i = 0; i < a->b.len; ++i) {
+        if (a->data[i] > b->data[i]) return false;
+        if (a->data[i] < b->data[i]) return true;
+        if (i >= b->b.len)  return false;
+    }
+    return a->b.len < b->b.len;
```

Comment:
It worked because the functions aren't symmetric. It's looping over the length of the left one always. Regardless, I'm going to switch to memcmp, which probably makes use of SSE.

Reply:
That is to say, if I'd always looped over the larger string, I could have made the code symmetric. I was trying to reduce the code as much as possible, but in the end I wager memcpy will be better.